src/_core/Storage.js

```diff
@@ -0,0 +1,70 @@
+class Storage {
+  constructor() {
+    this.timeout = undefined;
+    this.queuedChanges = {};
+    this.syncDelay = 3000;
+    this.initializeLocal(() => this.keepSynced());
+  };
+
+  initializeLocal(cb) {
+    chrome.storage.sync.get(null, (items) => {
+      chrome.storage.local.set(items, () => {
+        if (chrome.runtime.lastError) {
+          HFX.Logger.error(chrome.runtime.lastError.message);
+        }
+
+        cb();
+      });
+    });
+  };
+
+  keepSynced() {
+    if (chrome.storage.onChanged.hasListeners()) {
+      HFX.Logger.warn("Potentially browserifying the Storage module in multiple locations");
+      return;
+    }
+
+    chrome.storage.onChanged.addListener((changes, area) => {
+      if (changes._syncIgnore && changes._syncIgnore.newValue.area === area) {
+        return;
+      }
+
+      if (area === "local") {
+        if (this.timeout !== undefined) {
+          clearTimeout(this.timeout);
+        }
+
+        for (const key in changes) {
+          this.queuedChanges[key] = changes[key];
+        }
+
+        this.timeout = setTimeout(() => {
+          this.commitChanges("sync", chrome.storage.sync, this.queuedChanges);
+          this.queuedChanges = {};
+          this.timeout = undefined;
+        }, this.syncDelay);
+      } else {
+        this.commitChanges("local", chrome.storage.local, changes);
+      }
+    });
+  };
+
+  commitChanges(area, storage, changes) {
+    HFX.Logger.debug(`Committing changes to ${area}`);
+    const items = {};
+
+    for (const key in changes) {
+      items[key] = changes[key].newValue;
+    }
+
+    items._syncIgnore = { area: area, rand: Math.random() };
```

Comment:
Can you explain this bit to me?

Reply:
The `area` value is the area (`local` or `sync`) that is being synchronized.

So if `area` is `local`, then `local` is being updated to mirror `sync`. This change itself will fire an `onChange` event for `local`. By setting `area` to `local` this allows the `onChange` listener to know that the change was because of an update to `local` (the change was in `local` and `_syncIgnore.area` is `local`), so it doesn't need to synchronize those changes to `local` back to `sync` (since `sync` already has them). Otherwise, it would be continually synchronizing back and fourth.

The `rand` value is so that it will generate a change regardless of if `area` changed or not (in case there are consecutive synchronizations from one area to another).